multi/mcxn94x-multi/uart.c

```diff
@@ -0,0 +1,525 @@
+			break;
+	}
+
+	msgRespond(msg->oid.port, msg, rid);
+}
+
+
+static void uart_klogCblk(const char *data, size_t size)
+{
+	libtty_write(&uart_common.uarts[uartPos[uart_common.ttyminor]].tty_common, data, size, 0);
+}
+
+
+int uart_ttyInit(int ttyno)
+{
+	char path[24];
+	oid_t dev;
+
+	if (ttyno < 0) {
+		return -EINVAL;
+	}
+
+	snprintf(path, sizeof(path) - 1, "/dev/uart%d", ttyno);
+	path[sizeof(path) - 1] = '\0';
+
+	if (lookup(path, NULL, &dev) < 0) {
+		return -ENOENT;
+	}
+
+	uart_common.ttyminor = ttyno;
+
+	int err = symlink(path, _PATH_TTY);
+	if (err < 0) {
+		return err;
+	}
+
+	err = symlink(path, _PATH_CONSOLE);
+	if (err < 0) {
+		return err;
+	}
+
+	libklog_init(uart_klogCblk);
+
+	return 0;
+}
+
+
+static int uart_init(unsigned int minor)
+{
+	uint32_t t;
+	uart_t *uart;
+	libtty_callbacks_t callbacks;
+	static const size_t fifoSzLut[] = { 1, 4, 8, 16, 32, 64, 128, 256 };
+	static const struct {
+		volatile uint32_t *base;
+		int tx;
+		int rx;
+		int txalt;
+		int rxalt;
+		int irq;
+	} info[10] = {
+		{ .base = FLEXCOMM0_BASE, .tx = UART0_TX_PIN, .rx = UART0_RX_PIN, .txalt = UART0_TX_ALT, .rxalt = UART0_RX_ALT, .irq = FLEXCOMM0_IRQ },
+		{ .base = FLEXCOMM1_BASE, .tx = UART1_TX_PIN, .rx = UART1_RX_PIN, .txalt = UART1_TX_ALT, .rxalt = UART1_RX_ALT, .irq = FLEXCOMM1_IRQ },
+		{ .base = FLEXCOMM2_BASE, .tx = UART2_TX_PIN, .rx = UART2_RX_PIN, .txalt = UART2_TX_ALT, .rxalt = UART2_RX_ALT, .irq = FLEXCOMM2_IRQ },
+		{ .base = FLEXCOMM3_BASE, .tx = UART3_TX_PIN, .rx = UART3_RX_PIN, .txalt = UART3_TX_ALT, .rxalt = UART3_RX_ALT, .irq = FLEXCOMM3_IRQ },
+		{ .base = FLEXCOMM4_BASE, .tx = UART4_TX_PIN, .rx = UART4_RX_PIN, .txalt = UART4_TX_ALT, .rxalt = UART4_RX_ALT, .irq = FLEXCOMM4_IRQ },
+		{ .base = FLEXCOMM5_BASE, .tx = UART5_TX_PIN, .rx = UART5_RX_PIN, .txalt = UART5_TX_ALT, .rxalt = UART5_RX_ALT, .irq = FLEXCOMM5_IRQ },
+		{ .base = FLEXCOMM6_BASE, .tx = UART6_TX_PIN, .rx = UART6_RX_PIN, .txalt = UART6_TX_ALT, .rxalt = UART6_RX_ALT, .irq = FLEXCOMM6_IRQ },
+		{ .base = FLEXCOMM7_BASE, .tx = UART7_TX_PIN, .rx = UART7_RX_PIN, .txalt = UART7_TX_ALT, .rxalt = UART7_RX_ALT, .irq = FLEXCOMM7_IRQ },
+		{ .base = FLEXCOMM8_BASE, .tx = UART8_TX_PIN, .rx = UART8_RX_PIN, .txalt = UART8_TX_ALT, .rxalt = UART8_RX_ALT, .irq = FLEXCOMM8_IRQ },
+		{ .base = FLEXCOMM9_BASE, .tx = UART9_TX_PIN, .rx = UART9_RX_PIN, .txalt = UART9_TX_ALT, .rxalt = UART9_RX_ALT, .irq = FLEXCOMM9_IRQ },
+	};
+
+	static const uint32_t default_baud[] = { UART0_BAUDRATE, UART1_BAUDRATE, UART2_BAUDRATE, UART3_BAUDRATE,
+		UART4_BAUDRATE, UART5_BAUDRATE, UART6_BAUDRATE, UART7_BAUDRATE, UART8_BAUDRATE, UART9_BAUDRATE };
+	static const uint32_t tty_bufsz[] = { UART0_BUFFSZ, UART1_BUFFSZ, UART2_BUFFSZ, UART3_BUFFSZ,
+		UART4_BUFFSZ, UART5_BUFFSZ, UART6_BUFFSZ, UART7_BUFFSZ, UART8_BUFFSZ, UART9_BUFFSZ };
+
+	uart = &uart_common.uarts[uartPos[minor]];
+	uart->base = info[minor].base;
+
+	if ((condCreate(&uart->cond) < 0) || (mutexCreate(&uart->lock) < 0)) {
```

Comment:
If `mutexCreate` fails, we don't destroy the `cond` (does it also fall under MISRA 13.5 in this case? not sure)